src/Sema.zig

```diff
@@ -21154,356 +20673,602 @@ fn reifyEnum(
+    }
+
+    const ptr_size = try sema.interpretBuiltinType(block, operand_src, size_val, std.builtin.Type.Pointer.Size);
+
+    const actual_sentinel: InternPool.Index = s: {
+        if (!sentinel_val.isNull(zcu)) {
+            if (ptr_size == .one or ptr_size == .c) {
+                return sema.fail(block, src, "sentinels are only allowed on slices and unknown-length pointers", .{});
+            }
+            const sentinel_ptr_val = sentinel_val.optionalValue(zcu).?;
+            const ptr_ty = try pt.singleMutPtrType(elem_ty);
+            const sent_val = (try sema.pointerDeref(block, src, sentinel_ptr_val, ptr_ty)).?;
+            try sema.checkSentinelType(block, src, elem_ty);
+            if (sent_val.canMutateComptimeVarState(zcu)) {
+                const sentinel_name = try ip.getOrPutString(gpa, pt.tid, "sentinel_ptr", .no_embedded_nulls);
+                return sema.failWithContainsReferenceToComptimeVar(block, src, sentinel_name, "sentinel", sent_val);
+            }
+            break :s sent_val.toIntern();
+        }
+        break :s .none;
+    };
+
+    if (elem_ty.zigTypeTag(zcu) == .noreturn) {
+        return sema.fail(block, src, "pointer to noreturn not allowed", .{});
+    } else if (elem_ty.zigTypeTag(zcu) == .@"fn") {
+        if (ptr_size != .one) {
+            return sema.fail(block, src, "function pointers must be single pointers", .{});
+        }
+    } else if (ptr_size == .many and elem_ty.zigTypeTag(zcu) == .@"opaque") {
+        return sema.fail(block, src, "unknown-length pointer to opaque not allowed", .{});
+    } else if (ptr_size == .c) {
+        if (!try sema.validateExternType(elem_ty, .other)) {
+            const msg = msg: {
+                const msg = try sema.errMsg(src, "C pointers cannot point to non-C-ABI-compatible type '{f}'", .{elem_ty.fmt(pt)});
+                errdefer msg.destroy(gpa);
 
-    const loaded_union = ip.loadUnionType(wip_ty.index);
+                try sema.explainWhyTypeIsNotExtern(msg, src, elem_ty, .other);
 
-    const enum_tag_ty, const has_explicit_tag = if (opt_tag_type_val.optionalValue(zcu)) |tag_type_val| tag_ty: {
-        switch (ip.indexToKey(tag_type_val.toIntern())) {
-            .enum_type => {},
-            else => return sema.fail(block, src, "Type.Union.tag_type must be an enum type", .{}),
+                try sema.addDeclaredHereNote(msg, elem_ty);
+                break :msg msg;
+            };
+            return sema.failWithOwnedErrorMsg(block, msg);
         }
-        const enum_tag_ty = tag_type_val.toType();
-
-        // We simply track which fields of the tag type have been seen.
-        const tag_ty_fields_len = enum_tag_ty.enumFieldCount(zcu);
-        var seen_tags = try std.DynamicBitSetUnmanaged.initEmpty(sema.arena, tag_ty_fields_len);
+        if (elem_ty.zigTypeTag(zcu) == .@"opaque") {
+            return sema.fail(block, src, "C pointers cannot point to opaque types", .{});
+        }
```

Comment:
Note for when rebasing: this logic is gone on master, instead absorbed into the "unknown-length pointer to opaque..." check above (now written "indexable pointer to opaque...").